Release pgstream v0.4.0 (#356)
xataio/mdx-blog: whats-new-in-pgstream-0-4-0.mdx

---

> title: 'pgstream v0.4.0: Postgres-to-Postgres replication, snapshots & transformations'

Author: I think the title and the description say the same, thoughts on maybe changing the title to something simpler, like for pgroll updates?
@CezzaineZ @alexfrancoeur 
```suggestion
title: 'pgstream v0.4.0 update'
```

Author: As discussed https://xata-hq.slack.com/archives/C05474E6PRQ/p1743072481297669?thread_ts=1742927569.307849&cid=C05474E6PRQ, let's change the description instead.

> description: pgstream v0.4 includes Postgres to Postgres replication, snapshots and transformations.

Author: I think the title and the description say the same, thoughts on maybe changing the title to something simpler, like for pgroll updates?
`pgstream 0.4.0 update`
@CezzaineZ @alexfrancoeur

> description: pgstream v0.4 includes Postgres to Postgres replication, snapshots and transformations.

Author: ```suggestion
description: Learn how the latest features in pgstream refine Postgres replication with near real-time data capture, consistent snapshots, and column-level transformations.
```

image:
  src: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/pgstream/update-on-pgstream@2x.png
  alt: pgstream
author: Esther Minano Sanz
authorEmail: esther@xata.io
date: 03-27-2025
tags: ['open-source', 'postgres', 'schema', 'pgstream', 'replication', 'transformers', 'snapshots', 'oss']
published: true
slug: pgstream-0-4-0-update
ogImage: https://raw.githubusercontent.com/xataio/mdx-blog/main/images/pgstream/update-on-pgstream@2x.png
---

> We're proud to announce the release of `v0.4` of `[pgstream](https://github.com/xataio/pgstream)`, our open-source CDC tool for Postgres! 🚀 In this blog post, we'll dive into some of the key features packed into this latest release, and look at what the future holds!

Author: ```suggestion
We're proud to announce the release of `v0.4` of [`pgstream`](https://github.com/xataio/pgstream), our open-source CDC tool for Postgres! 🚀 In this blog post, we'll dive into some of the key features packed into this latest release, and look at what the future holds!
```

# Processor config
PGSTREAM_POSTGRES_WRITER_TARGET_URL="postgres://staging_user:password@staging-db:5432/staging"
PGSTREAM_POSTGRES_WRITER_BATCH_SIZE=100
PGSTREAM_POSTGRES_WRITER_BATCH_TIMEOUT=1s
```

Check out the [Postgres to Postgres tutorial](https://github.com/xataio/pgstream/blob/main/docs/tutorials/postgres_to_postgres.md#-postgresql-replication-to-postgresql-) for more details on how to set up and run replication with `pgstream`.

### Snapshots

![Snapshot to Postgres](/images/pgstream/pgstream_tutorial_snapshot2pg.svg)

With the addition of snapshots to `pgstream`, it's now possible to capture a consistent view of your Postgres database at a specific point in time, be it as an initial snapshot of the database before starting replication, or as a stand alone process when replication is not needed.

The snapshot listener produces the snapshot in different ways for the schema and the data in order to optimise the implementation for the different outputs:

- **Schema Snapshots**: Capture the database schema using the `pgstream.schema_log` table or Postgres utilities `pg_dump`/ `pg_restore` for Postgres to Postgres replication.

> - **Data Snapshots**: Use transaction snapshot IDs to obtain a stable view of the data, allowing to process tables and rows in parallel. When combined with replication, the current LSN is stored before starting the snapshot to ensure any WAL events that occur during the snapshot can be replayed.

Reviewer: ```suggestion
- **Data snapshots**: Use transaction snapshot IDs to obtain a stable view of the data, allowing to process tables and rows in parallel. When combined with replication, the current LSN is stored before starting the snapshot to ensure any WAL events that occur during the snapshot can be replayed.
```
